### PR TITLE
Add sorting based on edit distance to suggestions

### PR DESCRIPTION
**Pull request by HuwCampbell**

### CODE REVIEW

File: test/Icicle/Test/Internal/EditDistance.hs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  (===) 1 $ (editDistance (text $ as <> (pure a) <> bs) (text $ as <> (pure b) <> bs))
+
+prop_edit_flip_middle :: [Char] -> [Char] -> Char -> Char -> Property
+prop_edit_flip_middle as bs a b = a /= b ==>
+  (===) 1 $ (editDistance (text $ as <> (pure a) <> (pure b) <> bs) (text $ as <> (pure b) <> (pure a) <> bs))
+
+return []
+tests :: IO Bool
+-- tests = $quickCheckAll
+tests = $forAllProperties $ quickCheckWithResult (stdArgs { {- maxSuccess = 5000, -} maxSize = 10,  maxDiscardRatio = 10000})
```
Comment (amosr) [inline test/Icicle/Test/Internal/EditDistance.hs]: probably can just use standard $quickCheckAll; the maxDiscardRatio is only necessary for the "generate all queries and filter to only well-typed ones"

**Comment by amosr** [comment]
You'll probably need to run
```
test/cli/repl/run
# look to make sure only differences are in suggested error messages
test/cli/repl/run UPDATE
```
to update the repl test expected outputs


:ice_cream:

File: src/Icicle/Internal/EditDistance.hs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      a' = pretty_comp a
+      b' = pretty_comp b
+  in damerauLevenshtein a' b'
+
+-- This function it from the library language-spelling, but it has dependencies on
+-- ListLike, listlike instances and a version of Data.Text which clashes with what
+-- we're using. For just one function, it's not too terrible to just bring it into
+-- this file.
+
+damerauLevenshtein :: (Ord a) => [a] -> [a] -> Int
```
Comment (etorreborre) [inline src/Icicle/Internal/EditDistance.hs]: FTR Thomas Sutton has presented a nice way of implementing the [edit distance algorithm at FPSyd](https://speakerdeck.com/thsutton/dynamic-programming-in-haskell.) a few months ago. The generic implementation for dynamic programming problems fits on one slide (slide 13) and uses `Vector` to explore solutions (the code is available [here](https://github.com/thsutton/talk-dynamic-programming/tree/master/haskell) if you want to have a look at it).

**Comment by tranma** [comment]
nice :coffee:

File: test/Icicle/Test/Internal/EditDistance.hs (hunk 1 above)
Comment (HuwCampbell) [inline test/Icicle/Test/Internal/EditDistance.hs]: The rem_one test actually has quite a high discard ratio. It gave up on quickCheckAll.

Comment (amosr) [inline test/Icicle/Test/Internal/EditDistance.hs]: Ah, of course. My mistake

**Pull request merged**